### PR TITLE
add script to prepare arguments and run javadoc.exe

### PR DESCRIPTION
**Pull request by anmeng10101**

### CODE REVIEW

File: src/java2yaml/Common/Utility/FileUtility.cs [hunk 1]
```diff
@@ -24,7 +24,7 @@ public static IEnumerable<string> GetFilesByExtension(string rootPath, string ex
                 extension = "*";
             }
 
-            var pattern = string.Join("*." , extension);
+            var pattern = string.Concat("*.", extension);
```
Comment (superyyrrzz) [inline src/java2yaml/Common/Utility/FileUtility.cs]: personally, prefer `$"*.{extension}"`

File: src/java2yaml/Constants/Constants.cs [hunk 2]
```diff
@@ -14,23 +14,23 @@ public static class Constants
         public const string FolderName = "name";
 
         public const string Src = @"src\";
+        public const string Doc = @"_javadoc";
```
Comment (superyyrrzz) [inline src/java2yaml/Constants/Constants.cs]: No need to prepend `@`

File: src/java2yaml/Constants/Constants.cs [hunk 3]
```diff
@@ -14,23 +14,23 @@ public static class Constants
...
-        public static class BuildTool
+        public static class Extensions
         {
-            public const string Maven = "pom.xml";
+            public const string Jar = "jar";
+            public const string Java = "java";
         }
 
-        public static class JavadocArgumentModel
+        public static class BuildTool
```
Comment (superyyrrzz) [inline src/java2yaml/Constants/Constants.cs]: name: `BuildToolConfig`?

File: src/java2yaml/DataContracts/ConfigModel/JavadocOptionModel.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            option.OutputPath = sourcePath;
+
+            return option;
+        }
+
+        public static string ToCmdArguments(JavadocOptionModel option)
+        {
+            var optionBuilder = new StringBuilder();
+
+            foreach (var prop in typeof(JavadocOptionModel).GetProperties())
```
Comment (superyyrrzz) [inline src/java2yaml/DataContracts/ConfigModel/JavadocOptionModel.cs]: We can enumerate properties here. Usually we try avoiding reflection like `GetProperties`, as it is slow.

File: src/java2yaml/Steps/RunJavadoc.cs [hunk 5]
```diff
@@ -10,19 +13,145 @@ public RunJavadoc(string path)
...
+                Phase = StepName,
+                Level = LogLevel.Info,
+                Message = $" {count} java files to scan, detail in '{targetPath}'..."
+            });
+        }
+
+        // todo: fix doclet package to allow us invoke with this format "javadoc.exe @options @files"
+        private string GenerateOptions(string repositoryPath)
+        {
+            var option = new JavadocOptionModel();
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RunJavadoc.cs]: Can this be?
```
var option = JavadocOptionModel.Create(repositoryPath);
```
We can try put the logic inside `JavadocOptionModel.Create()`.

File: src/java2yaml/Steps/RunJavadoc.cs [hunk 6]
```diff
@@ -10,19 +13,145 @@ public RunJavadoc(string path)
...
+
+                ProcessUtility.Execute("javadoc.exe", commandLineArgs, RepositoryPath);
             }
             );
+        }
+
+        private void GenerateFileList(string repositoryPath)
+        {
+            int count = 0;
+            var inputPaths = GetInputPaths(repositoryPath).ToList();
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RunJavadoc.cs]: No need to `ToList()`.

File: src/java2yaml/Steps/RunJavadoc.cs [hunk 7]
```diff
@@ -10,19 +13,145 @@ public RunJavadoc(string path)
...
+            }
+            else
+            {
+                return RemoveExcludePaths(directories);
+            }
+        }
 
+        private List<string> GetInputPathsFromConfig(string repositoryPath)
+        {
+            return (_config.InputPaths
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RunJavadoc.cs]: minor:
```
             return _config.InputPaths
                .Where(p => p.StartsWith(repositoryPath))
                .ToList();
```

File: src/java2yaml/Steps/RunJavadoc.cs [hunk 8]
```diff
@@ -10,19 +13,145 @@ public RunJavadoc(string path)
...
 
+        private List<string> RemoveExcludePaths(List<string> directories)
+        {
+            var paths = new List<string>();
+
+            foreach (var dir in directories)
+            {
+                var subDirectories = Directory.GetDirectories(dir, "*.*", SearchOption.AllDirectories)
+                    .Where(p => !InExcludePaths(p))
+                    .Select(p => p)
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RunJavadoc.cs]: remove this line

File: src/java2yaml/Steps/RunJavadoc.cs [hunk 9]
```diff
@@ -10,19 +13,145 @@ public RunJavadoc(string path)
...
+                    .ToList();
+
+                paths.AddRange(subDirectories);
+            }
+
+            return paths;
+        }
+
+        private bool InExcludePaths(string dir)
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RunJavadoc.cs:141]: Can use `Any`: https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=netframework-4.8

File: src/java2yaml/Steps/RunJavadoc.cs [hunk 10]
```diff
@@ -10,19 +13,145 @@ public RunJavadoc(string path)
...
             {
-                //TODO
+                Guard.ArgumentNotNullOrEmpty(RepositoryPath, nameof(RepositoryPath));
+
+                _config = config;
+
+                GenerateFileList(RepositoryPath);
+
+                var options = GenerateOptions(RepositoryPath);
+                var commandLineArgs = string.Concat(options, " @files");
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RunJavadoc.cs]: Better to reference constants: `@"{options} @{Constants.Files}"`

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 9 above)
Comment (superyyrrzz) [reply src/java2yaml/Steps/RunJavadoc.cs:141]: `return _config.ExcludePaths.Any(p => dir.StartsWith(p));`

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 5 above)
Comment (superyyrrzz) [reply src/java2yaml/Steps/RunJavadoc.cs]:
```
new xxxModel() {
  xxx=xxx,
  xxx=xxx,
}
```

File: src/java2yaml/Common/Utility/FileUtility.cs (hunk 1 above)
Comment (anmeng10101) [reply src/java2yaml/Common/Utility/FileUtility.cs]: thanks a lot, great format, corrected

File: src/java2yaml/Constants/Constants.cs (hunk 2 above)
Comment (anmeng10101) [reply src/java2yaml/Constants/Constants.cs]: removed

File: src/java2yaml/Constants/Constants.cs (hunk 3 above)
Comment (anmeng10101) [reply src/java2yaml/Constants/Constants.cs]: changed

File: src/java2yaml/DataContracts/ConfigModel/JavadocOptionModel.cs (hunk 4 above)
Comment (anmeng10101) [reply src/java2yaml/DataContracts/ConfigModel/JavadocOptionModel.cs]: change to build argument string in explicit way 
``` 
``` 
return "  -classpath " + dependencies
	+ " -encoding UTF-8"
	+ " -docletpath " + docletPath
	+ " -doclet com.microsoft.doclet.DocFxDoclet"
	+ " -sourcepath " + sourcePath
	+ " -outputpath " + outputPath;
```

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 10 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RunJavadoc.cs]: changed to $"{options} @{Constants.Files}"

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 6 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RunJavadoc.cs]: removed

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 5 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RunJavadoc.cs]: removed this model

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 7 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RunJavadoc.cs]: corrected

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 8 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RunJavadoc.cs]: removed

File: src/java2yaml/Constants/Constants.cs [hunk 11]
```diff
@@ -14,23 +14,23 @@ public static class Constants
         public const string FolderName = "name";
 
         public const string Src = @"src\";
+        public const string Doc = "_javadoc";
         public const string PackageFolder = @"packages";
+
+        public const string DocletClass = "com.microsoft.doclet.DocFxDoclet";
         public const string DocletLocation = @"tools\docfx-doclet.jar";
+        public const string Files = @"files";
```
Comment (superyyrrzz) [inline src/java2yaml/Constants/Constants.cs]: `@` can also be removed in these simple strings.

**Review by superyyrrzz — APPROVED** [review]

**Pull request merged**